src/core/model_repository_manager.cc

```diff
@@ -589,47 +596,120 @@ Status
   for (auto& version_backend : it->second) {
-    std::lock_guard<std::recursive_mutex> lock(version_backend.second->mtx_);
-    if (versions.find(version_backend.first) != versions.end()) {
-      version_backend.second->repository_path_ = repository_path;
-      version_backend.second->model_config_ = model_config;
-      version_backend.second->next_action_ = ActionType::LOAD;
-    } else if (force_unload) {
-      version_backend.second->next_action_ = ActionType::UNLOAD;
+    auto version = version_backend.first;
+    BackendInfo* backend_info = (version_backend.second.second == nullptr)
+                                    ? version_backend.second.first.get()
+                                    : version_backend.second.second.get();
+
+    std::lock_guard<std::recursive_mutex> lock(backend_info->mtx_);
+    if (versions.find(version) != versions.end()) {
+      backend_info->repository_path_ = repository_path;
+      backend_info->model_config_ = model_config;
+      backend_info->next_action_ = ActionType::LOAD;
+      backend_info->platform_ = GetPlatform(model_config.platform());
+    } else if (!defer_unload) {
+      backend_info->next_action_ = ActionType::UNLOAD;
+    } else {
+      load_tracker->defer_unload_set_.emplace(version);
+      continue;
     }
 
-    auto version = version_backend.first;
-    auto backend_info = version_backend.second.get();
     // set version-wise callback before triggering next action
     if (OnComplete != nullptr) {
-      version_backend.second->OnComplete_ =
-          [version, backend_info, affected_version_cnt, OnComplete]() {
-            OnComplete(version, backend_info->state_, affected_version_cnt);
-          };
+      backend_info->OnComplete_ = [this, model_name, version, backend_info,
+                                   OnComplete, load_tracker]() {
+        {
+          std::lock_guard<std::recursive_mutex> map_lock(map_mtx_);
+          auto it = map_.find(model_name);
+
+          // Check if the version backend is loaded in background, if so,
+          // replace the current version backend and unload it.
+          auto vit = it->second.find(version);
+          if (vit->second.second.get() == backend_info) {
+            vit->second.second.swap(vit->second.first);
+            auto unload_backend = vit->second.second.get();
+            unloading_backends_[(uintptr_t)unload_backend] =
+                std::move(vit->second.second);
+            std::lock_guard<std::recursive_mutex> lock(unload_backend->mtx_);
+            unload_backend->next_action_ = ActionType::UNLOAD;
+            unload_backend->OnComplete_ = [this, unload_backend]() {
+              std::lock_guard<std::recursive_mutex> map_lock(map_mtx_);
+              unloading_backends_.erase((uintptr_t)unload_backend);
+            };
+            TriggerNextAction(model_name, version, unload_backend);
+          }
+
+          // Check if all versions are completed and perform deferred unload
```

Comment:
What happens if some or all of the loaded/loading model versions fail? Do we just cancel the unload? I think we should.

Reply:
I agree, maybe the unload should be cancelled only if all loading versions fail as the rollback of the whole load request

Reply:
I think it is more intuitive to abort the load unless all versions load correctly. Can you file a ticket.

Reply:
Created DLIS-1707. But one thing about aborting the load is that the reason of failure is not going to be shown in model index, so user will need to look at the server log to figure out what is the issue with the model.

Reply:
But the response to the load request will show the error, correct?

Reply:
True, but that just mentions what versions are not loaded, so we should add more detail to the error message?